Fix error messge for running a Flow app war in a servlet container in devmode

Currently if you build `beveragebuddy` war and try to run it in a servlet container in dev mode, it will fail.

This happens due to the fact that `DevModeInitializer` performs IO operations when started (tries to install npm packages etc.)
It should not do anything like this and either fail with a descriptive error ('start in production mode only' or similar) or assume that all resources needed for the app are in the war file and start normally. 

Example: 

```
docker run -it --rm --name tomcat -p 8080:8080 -v /Users/someonetoignore/Work/flow/flow-starters/beverage-starter-flow/target/beveragebuddy-2.0-SNAPSHOT.war:/usr/local/tomcat/webapps/beveragebuddy-2.0-SNAPSHOT.war tomcat
```

Produces the following error:

```
[localhost-startStop-1] INFO dev-updater - Creating a default ./package.json
[localhost-startStop-1] INFO dev-updater - Updating package.json and installing npm dependencies ...
 npm --no-package-lock install --save @polymer/iron-icon @polymer/iron-list @vaadin/vaadin-accordion @vaadin/vaadin-app-layout @vaadin/vaadin-button @vaadin/vaadin-checkbox @vaadin/vaadin-combo-box @vaadin/vaadin-context-menu @vaadin/vaadin-custom-field @vaadin/vaadin-date-picker @vaadin/vaadin-details @vaadin/vaadin-dialog @vaadin/vaadin-form-layout @vaadin/vaadin-grid @vaadin/vaadin-icons @vaadin/vaadin-item @vaadin/vaadin-list-box @vaadin/vaadin-login @vaadin/vaadin-lumo-styles @vaadin/vaadin-material-styles @vaadin/vaadin-notification @vaadin/vaadin-ordered-layout @vaadin/vaadin-progress-bar @vaadin/vaadin-radio-button @vaadin/vaadin-select @vaadin/vaadin-split-layout @vaadin/vaadin-tabs @vaadin/vaadin-text-field @vaadin/vaadin-time-picker @vaadin/vaadin-upload @webcomponents/webcomponentsjs
03-Apr-2019 06:32:11.695 SEVERE [localhost-startStop-1] org.apache.catalina.core.ContainerBase.addChildInternal ContainerBase.addChild: start:
 org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/beveragebuddy-2.0-SNAPSHOT]]
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:167)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:754)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:980)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1852)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.UncheckedIOException: java.io.IOException: Cannot run program "npm" (in directory "."): error=2, No such file or directory
	at com.vaadin.flow.server.frontend.NodeUpdatePackages.execute(NodeUpdatePackages.java:114)
	at com.vaadin.flow.server.startup.DevModeInitializer.onStartup(DevModeInitializer.java:109)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5225)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	... 10 more
Caused by: java.io.IOException: Cannot run program "npm" (in directory "."): error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at com.vaadin.flow.server.frontend.NodeUpdatePackages.updateDependencies(NodeUpdatePackages.java:218)
	at com.vaadin.flow.server.frontend.NodeUpdatePackages.updatePackageJsonDependencies(NodeUpdatePackages.java:174)
	at com.vaadin.flow.server.frontend.NodeUpdatePackages.execute(NodeUpdatePackages.java:107)
	... 13 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 16 more

03-Apr-2019 06:32:11.699 SEVERE [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployWAR Error deploying web application archive [/usr/local/tomcat/webapps/beveragebuddy-2.0-SNAPSHOT.war]
 java.lang.IllegalStateException: ContainerBase.addChild: start: org.apache.catalina.LifecycleException: Failed to start component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/beveragebuddy-2.0-SNAPSHOT]]
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:758)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:980)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1852)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

DevModeInitializer might be doing file system IO operations

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/NpmTemplateParser.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/NpmTemplateParser.java
@@ -153,10 +153,16 @@ public class NpmTemplateParser implements TemplateParser {
                 String port = service.getDeploymentConfiguration()
                         .getStringProperty(SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT, null);
                 if (port == null || port.isEmpty()) {
-                    throw new IllegalStateException("Unable to get webpack port via "
-                            + SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT + " property");
+                    statsUrl = service.getStaticResource("/" + stats);
+                    if (statsUrl == null) {
+                        throw new IllegalStateException(String.format(
+                                "Unable to get the stats file in app resources or through webpack. "
+                                        + "The webpack port is unavailable via %s property. ",
+                                SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT));
+                    }
+                } else {
+                    statsUrl = new URL("http://localhost:" + port + "/" + stats);
                 }
-                statsUrl = new URL("http://localhost:" + port + "/" + stats);
             }
 
             if (statsUrl != null) {

--- a/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/NpmTemplateParser.java
+++ b/flow-server/src/main/java/com/vaadin/flow/component/polymertemplate/NpmTemplateParser.java
@@ -144,25 +144,22 @@ public class NpmTemplateParser implements TemplateParser {
         if (content != null) {
             getLogger().debug("Found sources for the tag '{}' in the stats file '{}'", tag, stats);
         } else {
-            URL statsUrl;
-            if (service.getDeploymentConfiguration().isProductionMode()) {
-                // in production stats file is taken from the web resources
-                statsUrl = service.getStaticResource("/" + stats);
-            } else {
-                // in devmode stats file is taken from webpack via http
+            URL statsUrl = null;
+            if (!service.getDeploymentConfiguration().isProductionMode()) {
                 String port = service.getDeploymentConfiguration()
-                        .getStringProperty(SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT, null);
-                if (port == null || port.isEmpty()) {
-                    statsUrl = service.getStaticResource("/" + stats);
-                    if (statsUrl == null) {
-                        throw new IllegalStateException(String.format(
-                                "Unable to get the stats file in app resources or through webpack. "
-                                        + "The webpack port is unavailable via %s property. ",
-                                SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT));
-                    }
-                } else {
+                    .getStringProperty(SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT, null);
+                if (port != null && !port.isEmpty()) {
                     statsUrl = new URL("http://localhost:" + port + "/" + stats);
+                } else {
+                    getLogger().warn(
+                            "Unable to get the stats file through webpack. "
+                                    + "The webpack port is unavailable via '{}' property. ",
+                            SERVLET_PARAMETER_DEVMODE_WEBPACK_RUNNING_PORT);
                 }
+            }
+
+            if (statsUrl == null) {
+                statsUrl = service.getStaticResource("/" + stats);
             }
 
             if (statsUrl != null) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendToolsLocator.java
@@ -80,12 +80,14 @@ public class FrontendToolsLocator implements Serializable {
      * Verifies that the tool specified works by performing its test launch.
      *
      * @param toolPath
-     *            the path to a tool to check, not {@code null}
+     *            the path to a tool to check
      * @return {@code true} if the test launch had ended with successful error
      *         code, {@code false} otherwise
      */
     public boolean verifyTool(File toolPath) {
-        return executeCommand(toolPath.getAbsolutePath(), "-v")
+        return Optional.ofNullable(toolPath).filter(File::isFile)
+                .map(File::getAbsolutePath)
+                .flatMap(path -> executeCommand(path, "-v"))
                 .map(this::omitErrorResult).isPresent();
     }
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/DevModeInitializer.java
@@ -20,11 +20,13 @@ import javax.servlet.ServletContext;
 import javax.servlet.ServletException;
 import javax.servlet.ServletRegistration;
 import javax.servlet.annotation.HandlesTypes;
-
 import java.io.File;
 import java.io.Serializable;
 import java.util.Collection;
 import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.component.Component;
 import com.vaadin.flow.component.dependency.HtmlImport;
@@ -83,15 +85,28 @@ public class DevModeInitializer implements ServletContainerInitializer, Serializ
             return;
         }
 
-        AnnotationValuesExtractor extractor = new AnnotationValuesExtractor(new DefaultClassFinder(classes));
-        if (!config.getBooleanProperty(SERVLET_PARAMETER_DEVMODE_SKIP_UPDATE_NPM, false)) {
-            new NodeUpdatePackages(extractor).execute();
-        }
+        AnnotationValuesExtractor extractor = new AnnotationValuesExtractor(
+                new DefaultClassFinder(classes));
+        try {
+            if (!config.getBooleanProperty(
+                    SERVLET_PARAMETER_DEVMODE_SKIP_UPDATE_NPM, false)) {
+                new NodeUpdatePackages(extractor).execute();
+            }
 
-        if (!config.getBooleanProperty(SERVLET_PARAMETER_DEVMODE_SKIP_UPDATE_IMPORTS, false)) {
-            new NodeUpdateImports(extractor).execute();
-        }
+            if (!config.getBooleanProperty(
+                    SERVLET_PARAMETER_DEVMODE_SKIP_UPDATE_IMPORTS, false)) {
+                new NodeUpdateImports(extractor).execute();
+            }
 
-        DevModeHandler.start(config);
+            DevModeHandler.start(config);
+        } catch (Exception e) {
+            log().warn(
+                    "Failed to start a dev mode, hot reload is disabled. Continuing to start the application.",
+                    e);
+        }
+    }
+
+    private Logger log() {
+        return LoggerFactory.getLogger(getClass());
     }
  }

